modules/helper/EsxiServiceHelper.py

```diff
@@ -0,0 +1,84 @@
+from modules.configuration.Configuration import Configuration
+from interfaces.host import Host
+from modules.api.Atlas import Atlas
+from modules.api.SshHelper import SshHelper
+
+from threading import Thread
+from queue import Queue
+
+import logging
+
+
+logger = logging.getLogger('esxi')
+
+
+class EsxiServiceHelper:
+
+    def __init__(self, config: Configuration) -> None:
+        self.atlas = Atlas(config)
+        self.config = config
+
+        self._services = self.config.monitored_serivces
+        command_list = ["/etc/init.d/%s status" %
+                        service for service in self._services]
+        self._command = ' & '.join(command_list)
+
+        # Strange bug. Display name is different from registered service name
+        self._services = [item.replace('nsx-opsagent', 'opsAgent')
+                          for item in self._services]
+
+    def _worker(q: Queue, ssh_username: str, ssh_password: str,
+                command: str, services: list, output: list):
+        while not q.empty():
+            host: Host = q.get()
+
+            answer = SshHelper.execute_command(
+                host.address,
+                ssh_username,
+                ssh_password,
+                command
+            )
+
+            host.services = {}
+            if answer is None:
+                host.services = {k: False for k in services}
+                # Maybe add some info to host that something is wrong?
+                output.append(host)
+                continue
+
+            answer = answer.splitlines()
+            for service in services:
+                for line in answer:
+                    if service in line and 'is running' in line:
+                        host.services[service] = True
+                        # check next service
+                        break
+                    else:
+                        host.services[service] = False
+            output.append(host)
+
+    def get_all_service_stats(self) -> list:
+        # get hosts
+        hosts = self.atlas.get_esxi_hosts()
+
+        # get services
+        q = Queue()
+        [q.put(host) for host in hosts]
+
+        threads = []
+        output = []
+        for i in range(self.config.ssh_max_threads):
+            t = Thread(target=EsxiServiceHelper._worker, args=(
+                q,
+                self.config.esxi_user,
+                self.config.esxi_password,
+                self._command,
+                self._services,
+                output
+            ))
+            threads.append(t)
+            t.start()
+
+        [t.join() for t in threads]
```

Comment:
I don't get it. You open 10 threads for the same task? Where is the run variable used to control the loop? Could you please elaborate on this?

Reply:
Will add this to the code:
```
        # Explaination of applied multihreading approach:
        # Imagine 10 threads operating as workers
        # Each worker gets a shared queue with 'tasks' - in this case a list of hosts to process
        # All workers operate and take jobs(hosts) from the queue-stack until it is empty.
        # If empty - the workers will stop and the join releases.
        # All Python base types are considered thread-safe, so a queue should be safe too.
        # Since the threads are separated and communication is tricky - a reference to a
        # thread_safe output variable is passed, where the results will be stored in.
```

Reply:
Thanks for the description. In this case I would change the run variable. This is good practice so everyone understands right away that this is not used in the loop. 

Can you really make sure that every thread always terminates on an unsuccessful SSH connection, for example? Especially for the method of workers you created. Otherwise we have threads here working forever or stuck.

Reply:
Thats the job of the ssh module. It has timeout functionality and must abort if something is not working fine. If the threads wouldn't stop, the programm couldnt continue because of the join and actually nothing would be returned from the collector.

One problem towards the thread handling is that multiple requests at one time would cause many threads more in paralell. In normal case this should not happen.